File: src/charting/renderer/YAxisRendererHorizontalBarChart.java

package com.github.mikephil.charting.renderer;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Path;
import android.graphics.RectF;

import com.github.mikephil.charting.components.LimitLine;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.components.YAxis.AxisDependency;
import com.github.mikephil.charting.components.YAxis.YAxisLabelPosition;
import com.github.mikephil.charting.utils.MPPointD;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.Utils;
import com.github.mikephil.charting.utils.ViewPortHandler;

import java.util.List;

public class YAxisRendererHorizontalBarChart extends YAxisRenderer {

    public YAxisRendererHorizontalBarChart(viewPortHandler: ViewPortHandler, YAxis yAxis,
                                           Transformer trans) {
        super(viewPortHandler, yAxis, trans);

        this.mLimitLinePaint.setTextAlign(Align.LEFT);
    }

    /**
     * Computes the axis values.
     *
     * @param yMin - the minimum y-value in the data object for this axis
     * @param yMax - the maximum y-value in the data object for this axis
     */
    
    public computeAxis(let yMin, let yMax, boolean inverted) {

        // calculate the starting and entry polet of the y-labels (depending on
        // zoom / contentrect bounds)
        if (mViewPortHandler.contentHeight() > 10 && !mViewPortHandler.isFullyZoomedOutX()) {

            MPPointD p1 = this.mTrans.getValuesByTouchPoint(mViewPortHandler.contentLeft(),
                    this.mViewPortHandler.contentTop());
            MPPointD p2 = this.mTrans.getValuesByTouchPoint(mViewPortHandler.contentRight(),
                    this.mViewPortHandler.contentTop());

            if (!inverted) {
                yMin =  p1.x;
                yMax =  p2.x;
            } else {
                yMin =  p2.x;
                yMax =  p1.x;
            }

            MPPointD.recycleInstance(p1);
            MPPointD.recycleInstance(p2);
        }

        computeAxisValues(yMin, yMax);
    }

    /**
     * draws the y-axis labels to the screen
     */
    
    public renderAxisLabels(c: Canvas) {

        if (!mYAxis.isEnabled() || !mYAxis.isDrawLabelsEnabled())
            return;

        float[] positions = getTransformedPositions();

        this.mAxisLabelPaint.setTypeface(mYAxis.getTypeface());
        this.mAxisLabelPaint.setTextSize(mYAxis.getTextSize());
        this.mAxisLabelPaint.setColor(mYAxis.getTextColor());
        this.mAxisLabelPaint.setTextAlign(Align.CENTER);

        let baseYOffset = Utils.convertDpToPixel(2.5f);
        let textHeight = Utils.calcTextHeight(mAxisLabelPaint, "Q");

        AxisDependency dependency = this.mYAxis.getAxisDependency();
        YAxisLabelPosition labelPosition = this.mYAxis.getLabelPosition();

        let yPos = 0;

        if (dependency == AxisDependency.LEFT) {

            if (labelPosition == YAxisLabelPosition.OUTSIDE_CHART) {
                yPos = this.mViewPortHandler.contentTop() - baseYOffset;
            } else {
                yPos = this.mViewPortHandler.contentTop() - baseYOffset;
            }

        } else {

            if (labelPosition == YAxisLabelPosition.OUTSIDE_CHART) {
                yPos = this.mViewPortHandler.contentBottom() + textHeight + baseYOffset;
            } else {
                yPos = this.mViewPortHandler.contentBottom() + textHeight + baseYOffset;
            }
        }

        drawYLabels(c, yPos, positions, this.mYAxis.getYOffset());
    }

    
    public renderAxisLine(c: Canvas) {

        if (!mYAxis.isEnabled() || !mYAxis.isDrawAxisLineEnabled())
            return;

        this.mAxisLinePaint.setColor(mYAxis.getAxisLineColor());
        this.mAxisLinePaint.setStrokeWidth(mYAxis.getAxisLineWidth());

        if (mYAxis.getAxisDependency() == AxisDependency.LEFT) {
            c.drawLine(mViewPortHandler.contentLeft(),
                    this.mViewPortHandler.contentTop(), this.mViewPortHandler.contentRight(),
                    this.mViewPortHandler.contentTop(), this.mAxisLinePaint);
        } else {
            c.drawLine(mViewPortHandler.contentLeft(),
                    this.mViewPortHandler.contentBottom(), this.mViewPortHandler.contentRight(),
                    this.mViewPortHandler.contentBottom(), this.mAxisLinePaint);
        }
    }

    /**
     * draws the y-labels on the specified x-position
     *
     * @param fixedPosition
     * @param positions
     */
    
    protected drawYLabels(c: Canvas, let fixedPosition, float[] positions, let offset) {

        this.mAxisLabelPaint.setTypeface(mYAxis.getTypeface());
        this.mAxisLabelPaint.setTextSize(mYAxis.getTextSize());
        this.mAxisLabelPaint.setColor(mYAxis.getTextColor());

        const from = this.mYAxis.isDrawBottomYLabelEntryEnabled() ? 0 : 1;
        const to = this.mYAxis.isDrawTopYLabelEntryEnabled()
                ? this.mYAxis.mEntryCount
                : (mYAxis.mEntryCount - 1);

        for (let i = from; i < to; i++) {

            let text = this.mYAxis.getFormattedLabel(i);

            c.drawText(text, positions[i * 2], fixedPosition - offset, this.mAxisLabelPaint);
        }
    }

    
    protected float[] getTransformedPositions() {

        if(mGetTransformedPositionsBuffer.length != this.mYAxis.mEntryCount * 2) {
            this.mGetTransformedPositionsBuffer = new float[mYAxis.mEntryCount * 2];
        }
        float[] positions = this.mGetTransformedPositionsBuffer;

        for (let i = 0; i < positions.length; i += 2) {
            // only fill x values, y values are not needed for x-labels
            positions[i] = this.mYAxis.mEntries[i / 2];
        }

        this.mTrans.pointValuesToPixel(positions);
        return positions;
    }

    
    public RectF getGridClippingRect() {
        this.mGridClippingRect.set(mViewPortHandler.getContentRect());
        this.mGridClippingRect.inset(-mAxis.getGridLineWidth(), 0);
        return this.mGridClippingRect;
    }

    
    protected Path linePath(Path p, let i, float[] positions) {

        p.moveTo(positions[i], this.mViewPortHandler.contentTop());
        p.lineTo(positions[i], this.mViewPortHandler.contentBottom());

        return p;
    }

    protected Path this.mDrawZeroLinePathBuffer = new Path();

    
    protected drawZeroLine(c: Canvas) {

        let clipRestoreCount = c.save();
        this.mZeroLineClippingRect.set(mViewPortHandler.getContentRect());
        this.mZeroLineClippingRect.inset(-mYAxis.getZeroLineWidth(), 0);
        c.clipRect(mLimitLineClippingRect);

        // draw zero line
        MPPointD pos = this.mTrans.getPixelForValues(0, 0);

        this.mZeroLinePaint.setColor(mYAxis.getZeroLineColor());
        this.mZeroLinePaint.setStrokeWidth(mYAxis.getZeroLineWidth());

        Path zeroLinePath = this.mDrawZeroLinePathBuffer;
        zeroLinePath.reset();

        zeroLinePath.moveTo( pos.x - 1, this.mViewPortHandler.contentTop());
        zeroLinePath.lineTo( pos.x - 1, this.mViewPortHandler.contentBottom());

        // draw a path because lines don't support dashing on lower android versions
        c.drawPath(zeroLinePath, this.mZeroLinePaint);

        c.restoreToCount(clipRestoreCount);
    }

    protected Path this.mRenderLimitLinesPathBuffer = new Path();
    protected float[] this.mRenderLimitLinesBuffer = new float[4];
    /**
     * Draws the LimitLines associated with this axis to the screen.
     * This is the standard XAxis renderer using the YAxis limit lines.
     *
     * @param c
     */
    
    public renderLimitLines(c: Canvas) {

        List<LimitLine> limitLines = this.mYAxis.getLimitLines();

        if (limitLines == null || limitLines.length <= 0)
            return;

        float[] pts = this.mRenderLimitLinesBuffer;
        pts[0] = 0;
        pts[1] = 0;
        pts[2] = 0;
        pts[3] = 0;
        Path limitLinePath = this.mRenderLimitLinesPathBuffer;
        limitLinePath.reset();

        for (let i = 0; i < limitLines.length; i++) {

            LimitLine l = limitLines.get(i);

            if (!l.isEnabled())
                continue;

            let clipRestoreCount = c.save();
            this.mLimitLineClippingRect.set(mViewPortHandler.getContentRect());
            this.mLimitLineClippingRect.inset(-l.getLineWidth(), 0);
            c.clipRect(mLimitLineClippingRect);

            pts[0] = l.getLimit();
            pts[2] = l.getLimit();

            this.mTrans.pointValuesToPixel(pts);

            pts[1] = this.mViewPortHandler.contentTop();
            pts[3] = this.mViewPortHandler.contentBottom();

            limitLinePath.moveTo(pts[0], pts[1]);
            limitLinePath.lineTo(pts[2], pts[3]);

            this.mLimitLinePaint.setStyle(Paint.Style.STROKE);
            this.mLimitLinePaint.setColor(l.getLineColor());
            this.mLimitLinePaint.setPathEffect(l.getDashPathEffect());
            this.mLimitLinePaint.setStrokeWidth(l.getLineWidth());

            c.drawPath(limitLinePath, this.mLimitLinePaint);
            limitLinePath.reset();

            let label = l.getLabel();

            // if drawing the limit-value label is enabled
            if (label != null && !label.equals("")) {

                this.mLimitLinePaint.setStyle(l.getTextStyle());
                this.mLimitLinePaint.setPathEffect(null);
                this.mLimitLinePaint.setColor(l.getTextColor());
                this.mLimitLinePaint.setTypeface(l.getTypeface());
                this.mLimitLinePaint.setStrokeWidth(0.5f);
                this.mLimitLinePaint.setTextSize(l.getTextSize());

                let xOffset = l.getLineWidth() + l.getXOffset();
                let yOffset = Utils.convertDpToPixel(2f) + l.getYOffset();

                final LimitLine.LimitLabelPosition position = l.getLabelPosition();

                if (position == LimitLine.LimitLabelPosition.RIGHT_TOP) {

                    const labelLineHeight = Utils.calcTextHeight(mLimitLinePaint, label);
                    this.mLimitLinePaint.setTextAlign(Align.LEFT);
                    c.drawText(label, pts[0] + xOffset, this.mViewPortHandler.contentTop() + yOffset + labelLineHeight, this.mLimitLinePaint);
                } else if (position == LimitLine.LimitLabelPosition.RIGHT_BOTTOM) {

                    this.mLimitLinePaint.setTextAlign(Align.LEFT);
                    c.drawText(label, pts[0] + xOffset, this.mViewPortHandler.contentBottom() - yOffset, this.mLimitLinePaint);
                } else if (position == LimitLine.LimitLabelPosition.LEFT_TOP) {

                    this.mLimitLinePaint.setTextAlign(Align.RIGHT);
                    const labelLineHeight = Utils.calcTextHeight(mLimitLinePaint, label);
                    c.drawText(label, pts[0] - xOffset, this.mViewPortHandler.contentTop() + yOffset + labelLineHeight, this.mLimitLinePaint);
                } else {

                    this.mLimitLinePaint.setTextAlign(Align.RIGHT);
                    c.drawText(label, pts[0] - xOffset, this.mViewPortHandler.contentBottom() - yOffset, this.mLimitLinePaint);
                }
            }

            c.restoreToCount(clipRestoreCount);
        }
    }
}

File: src/charting/renderer/YAxisRendererHorizontalBarChart.ts
import { YAxisRenderer } from './YAxisRenderer';
import { ViewPortHandler } from '../utils/ViewPortHandler';
import { AxisDependency, YAxis, YAxisLabelPosition } from '../components/YAxis';
import { Transformer } from '../utils/Transformer';
import { Align, Canvas, Path, RectF, Style } from '@nativescript-community/ui-canvas';
import { Utils } from '../utils/Utils';
import { LimitLabelPosition } from '../components/LimitLine';
import { profile } from '@nativescript/core';

export class YAxisRendererHorizontalBarChart extends YAxisRenderer {
    /**
     * Computes the axis values.
     *
     * @param yMin - the minimum y-value in the data object for this axis
     * @param yMax - the maximum y-value in the data object for this axis
     */
    public computeAxis(yMin, yMax, inverted) {
        const axis = this.mYAxis;
        // calculate the starting and entry polet of the y-labels (depending on
        // zoom / contentrect bounds)
        const rect = axis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        if (rect.height() > 10 && !this.mViewPortHandler.isFullyZoomedOutX()) {
            const p1 = this.mTrans.getValuesByTouchPoint(rect.left, rect.top);
            const p2 = this.mTrans.getValuesByTouchPoint(rect.right, rect.top);

            if (!inverted) {
                yMin = p1.x;
                yMax = p2.x;
            } else {
                yMin = p2.x;
                yMax = p1.x;
            }
        }

        this.computeAxisValues(yMin, yMax);
    }

    /**
     * draws the y-axis labels to the screen
     */
    @profile
    public renderAxisLabels(c: Canvas) {
        const axis = this.mYAxis;
        if (!axis.isEnabled() || !axis.isDrawLabelsEnabled()) {
            return;
        }

        const positions = this.getTransformedPositions();
        const paint = this.axisLabelsPaint;
        paint.setFont(axis.getFont());
        paint.setColor(axis.getTextColor());
        paint.setTextAlign(Align.CENTER);

        const baseYOffset = 2.5;
        const textHeight = Utils.calcTextHeight(paint, 'Q');

        const dependency = axis.getAxisDependency();
        const labelPosition = axis.getLabelPosition();

        let yPos = 0;

        const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        if (dependency === AxisDependency.LEFT) {
            if (labelPosition === YAxisLabelPosition.OUTSIDE_CHART) {
                yPos = rect.top - baseYOffset;
            } else {
                yPos = rect.top - baseYOffset;
            }
        } else {
            if (labelPosition === YAxisLabelPosition.OUTSIDE_CHART) {
                yPos = rect.bottom + textHeight + baseYOffset;
            } else {
                yPos = rect.bottom + textHeight + baseYOffset;
            }
        }

        this.drawYLabels(c, yPos, positions, axis.getYOffset());
    }

    public renderAxisLine(c: Canvas) {
        const axis = this.mYAxis;
        if (!axis.isEnabled() || !axis.isDrawAxisLineEnabled()) {
            return;
        }

        const paint = this.axisLinePaint;
        paint.setColor(axis.getAxisLineColor());
        paint.setStrokeWidth(axis.getAxisLineWidth());

        const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        if (axis.getAxisDependency() === AxisDependency.LEFT) {
            c.drawLine(rect.left, rect.top, rect.right, rect.top, paint);
        } else {
            c.drawLine(rect.left, rect.bottom, rect.right, rect.bottom, paint);
        }
    }

    /**
     * draws the y-labels on the specified x-position
     *
     * @param fixedPosition
     * @param positions
     */
    @profile
    protected drawYLabels(c: Canvas, fixedPosition, positions, offset) {
        const axis = this.mYAxis;
        const paint = this.axisLabelsPaint;
        paint.setFont(axis.getFont());
        paint.setColor(axis.getTextColor());

        const from = axis.isDrawBottomYLabelEntryEnabled() ? 0 : 1;
        const to = axis.isDrawTopYLabelEntryEnabled() ? axis.mEntryCount : axis.mEntryCount - 1;

        for (let i = from; i < to; i++) {
            const text = axis.getFormattedLabel(i);
            if (text) {
                c.drawText(text + '', positions[i * 2], fixedPosition - offset, paint);
            }
        }
    }

    protected getTransformedPositions() {
        const axis = this.mYAxis;
        const length = axis.mEntryCount * 2;
        if (!this.mGetTransformedPositionsBuffer || this.mGetTransformedPositionsBuffer.length !== length) {
            this.mGetTransformedPositionsBuffer = Utils.createArrayBuffer(length);
        }

        const positions = this.mGetTransformedPositionsBuffer;
        for (let i = 0; i < length; i += 2) {
            // only fill x values, y values are not needed for x-labels
            positions[i] = axis.mEntries[i / 2];
        }

        const result = Utils.pointsFromBuffer(positions);
        this.mTrans.pointValuesToPixel(result);
        return Utils.nativeArrayToArray(result);
    }

    public getGridClippingRect(): RectF {
        const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        const temRect = Utils.getTempRectF();
        temRect.set(rect);
        temRect.inset(-this.mAxis.getGridLineWidth(), 0);
        return temRect;
    }

    protected drawZeroLine(c: Canvas) {
        const axis = this.mYAxis;
        const clipRestoreCount = c.save();
        const rect = this.mAxis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        const zeroLineClippingRect = Utils.getTempRectF();
        zeroLineClippingRect.set(rect);
        zeroLineClippingRect.inset(-axis.getZeroLineWidth(), 0);
        c.clipRect(zeroLineClippingRect);

        // draw zero line
        const pos = this.mTrans.getPixelForValues(0, 0);
        const paint = this.zeroLinePaint;

        paint.setColor(axis.getZeroLineColor());
        paint.setStrokeWidth(axis.getZeroLineWidth());

        const zeroLinePath = Utils.getTempPath();
        zeroLinePath.reset();

        zeroLinePath.moveTo(pos.x - 1, rect.top);
        zeroLinePath.lineTo(pos.x - 1, rect.bottom);

        // draw a path because lines don't support dashing on lower android versions
        c.drawPath(zeroLinePath, paint);

        c.restoreToCount(clipRestoreCount);
    }

    /**
     * Draws the LimitLines associated with this axis to the screen.
     * This is the standard XAxis renderer using the YAxis limit lines.
     *
     * @param c
     */

    public renderLimitLines(c: Canvas) {
        const axis = this.mYAxis;
        const limitLines = axis.getLimitLines();

        if (limitLines == null || limitLines.length <= 0) {
            return;
        }

        const pts = Utils.getTempArray(2);
        pts[0] = 0;
        pts[1] = 0;
        // const limitLinePath = this.renderLimitLinesPath;
        // limitLinePath.reset();

        const customRender = axis.getCustomRenderer();
        const customRendererFunc = customRender && customRender.drawLimitLine;
        const rect = axis.isIgnoringOffsets() ? this.mViewPortHandler.getChartRect() : this.mViewPortHandler.getContentRect();
        const clipToContent = axis.clipLimitLinesToContent;
        for (let i = 0; i < limitLines.length; i++) {
            const l = limitLines[i];

            if (!l.isEnabled()) {
                continue;
            }

            const lineWidth = l.getLineWidth();
            if (clipToContent) {
                c.save();
                const clipRect = Utils.getTempRectF();
                clipRect.set(rect);
                clipRect.inset(0, -lineWidth);
                c.clipRect(clipRect);
            }

            pts[0] = l.getLimit();

            this.mTrans.pointValuesToPixel(pts);

            const paint = this.limitLinePaint;
            paint.setColor(l.getLineColor());
            paint.setPathEffect(l.getDashPathEffect());
            paint.setStrokeWidth(l.getLineWidth());

            if (lineWidth > 0) {
                if (customRendererFunc) {
                    customRendererFunc(c, this, l, rect, pts[0], paint);
                } else {
                    c.drawLine(pts[0], rect.bottom, pts[0], rect.top, paint);
                }
            }

            const label = l.getLabel();

            // if drawing the limit-value label is enabled
            if (label != null && label !== '') {
                paint.setStyle(l.getTextStyle());
                paint.setPathEffect(null);
                paint.setColor(l.getTextColor());
                paint.setFont(l.getFont());
                paint.setStrokeWidth(0.5);
                paint.setFont(l.getFont());

                const xOffset = l.getLineWidth() + l.getXOffset();
                const yOffset = 2 + l.getYOffset();

                const position = l.getLabelPosition();

                switch (position) {
                    case LimitLabelPosition.RIGHT_TOP: {
                        const labelLineHeight = Utils.calcTextHeight(paint, label);
                        paint.setTextAlign(Align.LEFT);
                        c.drawText(label, pts[0] + xOffset, rect.top + yOffset + labelLineHeight, paint);
                        break;
                    }
                    case LimitLabelPosition.RIGHT_BOTTOM: {
                        paint.setTextAlign(Align.LEFT);
                        c.drawText(label, pts[0] + xOffset, rect.bottom - yOffset, paint);
                        break;
                    }
                    case LimitLabelPosition.CENTER_TOP: {
                        const labelLineHeight = Utils.calcTextHeight(paint, label);
                        paint.setTextAlign(Align.CENTER);
                        c.drawText(label, pts[0], rect.top + yOffset + labelLineHeight, paint);
                        break;
                    }
                    case LimitLabelPosition.CENTER_BOTTOM: {
                        paint.setTextAlign(Align.CENTER);
                        c.drawText(label, pts[0], rect.bottom - yOffset, paint);
                        break;
                    }
                    case LimitLabelPosition.LEFT_TOP: {
                        paint.setTextAlign(Align.RIGHT);
                        const labelLineHeight = Utils.calcTextHeight(paint, label);
                        c.drawText(label, pts[0] - xOffset, rect.top + yOffset + labelLineHeight, paint);
                        break;
                    }
                    default: {
                        paint.setTextAlign(Align.RIGHT);
                        c.drawText(label, pts[0] - xOffset, rect.bottom - yOffset, paint);
                        break;
                    }
                }
            }

            if (clipToContent) {
                c.restore();
            }
        }
    }
}
